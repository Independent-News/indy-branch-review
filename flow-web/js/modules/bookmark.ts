import { COOKIE_LOGGED_IN } from '#app/constants/cookies';

import {
  CLASS_BOOKMARK_BUTTON,
  CLASS_BOOKMARK_NOTIFICATION,
  CLASS_BOOKMARK_NOTIFICATION_CLOSE,
  CLASS_BOOKMARK_NOTIFICATION_SHOW,
  CLASS_BOOKMARK_NOTIFICATION_LOGIN,
} from '#app/constants/classNames';
import { BOOKMARK_OVERLAY_LOGIN_RSM } from '#app/constants/regSourceMethods';

import { InternalApi } from '../utils/internalApi';

import * as thisModule from './bookmark';
import { getCookie } from './cookie';
import { dispatchArticleBookmark } from './custom-event-dispatchers';
import { openLoginDrawer } from './loginDrawer';

const initNotification = () => {
  const notification = document.querySelector(
    `.${CLASS_BOOKMARK_NOTIFICATION}`,
  );
  if (!notification) {
    return;
  }

  const closeNotification = notification.querySelector(
    `.${CLASS_BOOKMARK_NOTIFICATION_CLOSE}`,
  );
  if (closeNotification) {
    closeNotification.addEventListener('click', () => {
      notification.classList.remove(CLASS_BOOKMARK_NOTIFICATION_SHOW);
    });
  }

  const ignoreCheckbox = notification.querySelector('input');
  if (ignoreCheckbox) {
    ignoreCheckbox.addEventListener('click', () => {
      localStorage.setItem('ignore-bookmark', 'true');
    });
  }
};

const initLightbox = () => {
  if (!window.JSGlobals.featureFlags.feat__no_amp) {
    const loginBtns = [
      ...document.querySelectorAll(`.${CLASS_BOOKMARK_NOTIFICATION_LOGIN}`),
    ];
    loginBtns.forEach((btn) => {
      btn.addEventListener('click', () => {
        const closeButton = document.querySelector('#bookmark-lightbox .close');
        if (closeButton instanceof HTMLButtonElement) {
          closeButton.click();
        }

        setTimeout(() => {
          openLoginDrawer(BOOKMARK_OVERLAY_LOGIN_RSM);
        }, 400);
      });
    });
  }
};

const showBookmarkNotification = () => {
  const notification = document.querySelector('.bookmark-notification');
  const ignoreNotification = localStorage.getItem('ignore-bookmark');
  if (!notification || ignoreNotification) {
    return;
  }

  notification.classList.add(CLASS_BOOKMARK_NOTIFICATION_SHOW);
  setTimeout(() => {
    notification.classList.remove(CLASS_BOOKMARK_NOTIFICATION_SHOW);
  }, 3000);
};

const addBookmarkLocalStorage = (articleId: string) => {
  const bookmarks = JSON.parse(
    localStorage.getItem('premiumBookmarks') || '[]',
  );
  bookmarks.push(articleId);
  localStorage.setItem('premiumBookmarks', JSON.stringify(bookmarks));
};

export const removeBookmarkLocalStorage = (articleId: string) => {
  let bookmarks = JSON.parse(localStorage.getItem('premiumBookmarks') || '[]');

  bookmarks = bookmarks.filter((a: string) => a !== articleId);
  localStorage.setItem('premiumBookmarks', JSON.stringify(bookmarks));
};

export const deleteBookmark = async (
  articleId: string,
  button: HTMLButtonElement,
) => {
  if (button.classList.contains('submitting')) {
    return;
  }
  try {
    button.classList.add('submitting');
    const res = await InternalApi.delete(`bookmarks/${articleId}`);
    if (res.ok) {
      [...document.querySelectorAll(`.${CLASS_BOOKMARK_BUTTON}`)].forEach(
        (btn) => btn.classList.remove('saved'),
      );
      button.classList.remove('submitting');
      thisModule.removeBookmarkLocalStorage(articleId);
      return;
    }
    throw new Error('Network response was not ok.');
  } catch (error) {
    [...document.querySelectorAll(`.${CLASS_BOOKMARK_BUTTON}`)].forEach((btn) =>
      btn.classList.add('saved'),
    );
    button.classList.remove('submitting');
    console.error('Fetch failed: ', error.message);
  }
};

export const submitBookmark = async (
  articleId: string,
  button: HTMLButtonElement,
) => {
  if (button.classList.contains('submitting')) {
    return;
  }
  button.classList.add('submitting');
  const data = { articleId };
  InternalApi.post('bookmarks', data)
    .then((response) => {
      if (response.ok) {
        return response.json();
      }
      throw new Error('Network response was not ok.');
    })
    .then((response) => {
      if (response.saved) {
        [...document.querySelectorAll(`.${CLASS_BOOKMARK_BUTTON}`)].forEach(
          (btn) => btn.classList.add('saved'),
        );
        addBookmarkLocalStorage(articleId);
        showBookmarkNotification();
        dispatchArticleBookmark();
      } else {
        [...document.querySelectorAll(`.${CLASS_BOOKMARK_BUTTON}`)].forEach(
          (btn) => btn.classList.remove('saved'),
        );
        removeBookmarkLocalStorage(articleId);
      }
      button.classList.remove('submitting');
    })
    .catch((error) => {
      button.classList.remove('submitting');
      console.error('Fetch failed: ', error.message);
    });
};

export const updateBookmarkState = (
  articleId: string,
  button: HTMLButtonElement,
) => {
  const callback = button.classList.contains('saved')
    ? thisModule.deleteBookmark
    : thisModule.submitBookmark;
  callback(articleId, button);
};

const checkSavedBookmark = (articleId: string, button: HTMLButtonElement) => {
  const bookmarks = JSON.parse(
    localStorage.getItem('premiumBookmarks') || '[]',
  );

  if (bookmarks.indexOf(articleId) > -1) {
    button.classList.add('saved');
  } else {
    button.classList.remove('saved');
  }
};

const initBookmark = async () => {
  const bookmarkBtns = [
    ...document.querySelectorAll(`.${CLASS_BOOKMARK_BUTTON}`),
  ];
  const bookmarkNotification = document.querySelector(
    `.${CLASS_BOOKMARK_NOTIFICATION}`,
  );

  if (!bookmarkNotification) {
    return;
  }

  bookmarkBtns.forEach((btn) => {
    if (btn instanceof HTMLButtonElement) {
      const articleId = btn.dataset.articleId || '';
      if (getCookie(COOKIE_LOGGED_IN)) {
        checkSavedBookmark(articleId, btn);
      }

      btn.addEventListener('click', () => {
        updateBookmarkState(articleId, btn);
      });
    }
  });
  initLightbox();
  initNotification();
};

export default initBookmark;
