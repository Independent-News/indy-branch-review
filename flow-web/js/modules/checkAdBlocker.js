import { dispatchAdBlockDetected } from './custom-event-dispatchers';

export default async () => {
  // try to load dotmetrics - if this is blocked we know we have an ad-blocker running
  try {
    await fetch('https://www.npttech.com/advertising.js');
  } catch {
    dispatchAdBlockDetected();
    return true;
  }
  return false;
};
