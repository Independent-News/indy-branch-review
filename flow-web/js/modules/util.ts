/* globals */

import { escapeRegExp } from '#app/util/regex';

type ScriptAttributes = Array<[string, string]>;

export const getParameter = (name: string, url?: string) => {
  const urlInstance = new URL(url || window.location.href);

  return urlInstance.searchParams.get(name);
};

export const getParametersByNames = <T extends string>(
  names: Array<T> = [],
  url?: string,
): Array<{ key: T; value: string }> => {
  const urlInstance = new URL(url || window.location.href);

  return names.map((name) => {
    const paramValue = urlInstance.searchParams.get(name) ?? '';

    return {
      key: name,
      value: paramValue,
    };
  });
};

export const hasParameter = (name: string, url = null) => {
  const re = new RegExp(
    `[\\?|\\&]${escapeRegExp(encodeURIComponent(name))}\\b`,
  );
  return re.test(url || window.location.href);
};

export const getWindowWidth = () =>
  window.innerWidth ||
  document.documentElement.clientWidth ||
  document.body.clientWidth;

const createScriptElement = (
  url: string,
  {
    id,
    async,
    defer,
    ok,
    error,
    attributes,
  }: {
    id?: string;
    async?: boolean;
    defer?: boolean;
    ok: (value: unknown) => void;
    error: () => void;
    attributes?: ScriptAttributes;
  },
) => {
  const elem = document.createElement('script');

  elem.onload = ok;
  elem.onerror = error;

  if (id && id !== 'undefined') {
    elem.id = id;
  }

  if (async && !defer) {
    elem.setAttribute('async', '');
  }

  if (defer) {
    elem.setAttribute('defer', '');
  }

  if (attributes) {
    attributes.forEach(([key, value]) => {
      elem.setAttribute(key, value);
    });
  }

  elem.src = url;

  return elem;
};

export const loadJS = (
  urls: string | Array<string>,
  {
    async = true,
    defer = document.location.search.includes('___FORCE_DEFER___'),
    parent = document.body,
    attributes,
  }: {
    async?: boolean;
    defer?: boolean;
    parent?: HTMLElement;
    attributes?: ScriptAttributes;
  } = {},
) => {
  window.JSGlobals = window.JSGlobals || {};
  const loadedJS: Array<string> = (window.JSGlobals.loadedJS = []);

  return new Promise((resolve, reject) => {
    const promises = (urls instanceof Array ? urls : [urls]).map((url) =>
      loadedJS.includes(url)
        ? Promise.resolve()
        : new Promise((ok, error) => {
            const elem = createScriptElement(url, {
              async,
              defer,
              ok,
              error,
              attributes,
            });
            parent.appendChild(elem);
            loadedJS.push(url);
          }),
    );
    return Promise.all(promises).then(resolve, reject);
  });
};

export const loadCSS = (url: string) =>
  new Promise((resolve, reject) => {
    const elem = document.createElement('link');

    elem.onload = resolve;
    elem.onerror = reject;

    elem.rel = 'stylesheet';
    elem.href = url;

    document.head.insertAdjacentElement('afterbegin', elem);
  });

export const hasSpanishBrowserLanguage = (languages: Array<string>) => {
  for (let i = 1; i < languages.length; i++) {
    if (languages[i].includes('es')) {
      return true;
    }
  }
  return false;
};

/**
 * Locks the body from scrolling when the modal is open
 * @param {boolean} [toggle=true] Flag to control behaviour
 */
export const toggleBodyScroll = (toggle = true) => {
  document.body.classList.toggle('is-scroll-locked', toggle);
};

/**
 * Storing attribute values to prevent code duplication for cookie sync iframes
 */
export const cookieSyncIframeAttributes = {
  title: 'User Sync',
  width: 0,
  height: 0,
  sandbox: 'allow-same-origin allow-scripts',
  frameborder: 0,
  style: 'display: none;',
};

/**
 * Injects iframe to body
 * @param {object} attributes Object containing iframe attribute values
 */
export const injectIframe = (attributes: Record<string, string>) => {
  const iframe = document.createElement('iframe');

  Object.entries(attributes).forEach(([attribute, value]) => {
    iframe.setAttribute(attribute, value);
  });

  document.body.appendChild(iframe);
};
