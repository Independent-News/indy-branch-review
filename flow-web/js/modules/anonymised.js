import { checkVendorConsent } from './cmp';
import { loadJS } from './util';

const SP_VENDOR_ID = '6274ee2d293cdf1bf415179b';

export default async () => {
  if (await checkVendorConsent(SP_VENDOR_ID, 'anonymised')) {
    await loadJS('//static.anonymised.io/light/loader.js', {
      id: 'idward-plugin-js',
      attributes: [
        ['idw_client_id', 'NDEx'],
        ['idw_hide_button', 'true'],
      ],
    });
  }
};
