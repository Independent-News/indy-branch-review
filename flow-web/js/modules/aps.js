import { PUB_ID } from '#app/constants/aps';

import * as thisModule from './aps';
import { loadJS } from './util';

/**
 * Load Amazon Publish Service script
 */
export const loadScript = async () => {
  await loadJS(['/js/third-party/aps.js']);
};

/**
 * Initialize Amazon Publish Service
 */
export const init = () => {
  window.apstag?.init({
    pubID: PUB_ID,
    adServer: 'googletag',
    videoAdServer: 'DFP',
    deals: true,
  });
};

/**
 * @typedef {object} APSBid
 * @param {string} slotID The div ID of the slot that will render the creative
 * @param {string} mediaType The media type of the bid
 * @param {Array<string, string>} targeting The targeting key-value pairs for the bid
 */

/**
 * Get bids data from Amazon Publish Service bids request
 * @param {string} adUnitCode the ad unit code
 * @returns {Promise<Array<APSBid>>}
 * @see {@link https://ams.amazon.com/webpublisher/uam/docs/web-integration-documentation/integration-guide/javascript-guide/api-reference.html#apstagfetchbids}
 */
export const requestBids = (adUnitCode) => {
  const APS_TIMEOUT = 5000;
  // Directly added window to prevent AMP error
  const apstag = window.apstag;

  thisModule.init();

  const bidConfig = {
    slots: [
      {
        slotID: adUnitCode,
        mediaType: 'video',
      },
    ],
    timeout: APS_TIMEOUT,
  };

  return new Promise(function (resolve) {
    /**
     * Callback function required by apstag.fetchBids
     * @param {Array<APSBid>} bids
     */
    const callback = (bids) => {
      resolve(bids);
    };

    apstag.fetchBids(bidConfig, callback);
  });
};

/**
 * Get targeting key-value pairs from Amazon Publish Service bids request
 * @param {string} adUnitCode The ad unit code used as slot ID in the request
 * @returns {Array} targeting key-value pairs
 */
export const getTargeting = async (adUnitCode) => {
  const bids = await thisModule.requestBids(adUnitCode);
  const bid = bids.filter((bid) => bid.slotID === adUnitCode)[0];
  return bid?.targeting || [];
};
