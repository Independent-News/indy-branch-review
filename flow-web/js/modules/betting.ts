/**
 * Loads the analytics script based on consent and environment.
 *
 * Currently, we have two distinct JavaScript files:
 * - `https://adtech.indyadvertising.com/js/app.js` for betting offers which is
 * implemented in the google tag manager.
 * - `https://adtech.indyadvertising.com/js/ktag.js` for performance/revenue
 * tracking.
 *
 * To prevent cookies from dropping before consent, we need to add the 'ktag'
 * JS file to our code so it only fires when consent is given.
 *
 * The script URL varies depending on the environment:
 * - **Production**: `https://adtech.indyadvertising.com/js/ktag.js`
 * - **Development**: `https://adtech.indyadvertising.com/js/ktag-testing.js`
 * @param {boolean} [consent=false] - Whether the user has given consent
 * @returns {void}
 */

import { loadJS } from './util';

const bettingAnalyticsTag = (consent = false) => {
  const { url } = window.JSGlobals?.betting?.analytics || {};

  if (!url) {
    console.error('No betting analytics tag URL', {
      cause: window.JSGlobals?.betting,
    });
  }

  if (consent && url) {
    // 5s delay before loading the snippet is required
    window.setTimeout(() => {
      loadJS(url);
    }, 5000);
  }
};

export default bettingAnalyticsTag;
