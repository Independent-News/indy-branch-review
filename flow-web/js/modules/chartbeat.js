/* globals JSGlobals */

import { EVENT_CHARTBEAT_LOADED } from '#app/constants/customEvents';

import { loadJS } from './util';

export default async () => {
  if (!JSGlobals.chartbeat) {
    return;
  } // do nothing if no chartbeat data found on page.

  const digitalData = (window.digitalData = window.digitalData || {});

  /** CONFIGURATION START **/
  const _sf_async_config = (window._sf_async_config =
    window._sf_async_config || {});
  _sf_async_config.idSync = _sf_async_config.idSync || {};

  _sf_async_config.flickerControl = false;
  _sf_async_config.useCanonical = true;
  _sf_async_config.useCanonicalDomain = true;
  _sf_async_config.autoDetect = false; // Disable auto-tracking of all video players

  _sf_async_config.uid = JSGlobals.chartbeat.uid;
  _sf_async_config.title = JSGlobals.chartbeat.title;
  _sf_async_config.authors = JSGlobals.chartbeat.authors;
  _sf_async_config.sections = JSGlobals.chartbeat.sections;
  _sf_async_config.idSync.l = JSGlobals.userUid;
  _sf_async_config.idSync.hashed_email = JSGlobals.hashedEmail;

  const _cbq = (window._cbq = window._cbq || []);
  if (digitalData.user_status === 'premium') {
    _cbq.push(['_acct', 'paid']);
  } else if (digitalData.user_status === 'registered') {
    _cbq.push(['_acct', 'lgdin']);
  } else {
    _cbq.push(['_acct', 'anon']);
  }

  /** CONFIGURATION END **/
  await loadJS([
    '//static.chartbeat.com/js/chartbeat_video.js',
    '//static.chartbeat.com/js/chartbeat_mab.js',
  ]);

  document.dispatchEvent(new CustomEvent(EVENT_CHARTBEAT_LOADED));
};

export const trackVideo = (player) => {
  const _cbv = window._cbv || (window._cbv = []);
  _cbv.push(player);
};
