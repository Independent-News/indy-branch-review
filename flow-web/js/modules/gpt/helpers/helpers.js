import { getWindowWidth } from '../../util';

import * as thisModule from './helpers';

window.googletag = window.googletag || { cmd: [] };

export const createAd = (id, size, adUnitPath) => {
  const conditionalAdElem = document.createElement('div');
  conditionalAdElem.id = id;
  conditionalAdElem.setAttribute('data-mpu', 'true');
  conditionalAdElem.setAttribute('data-size', size);
  conditionalAdElem.setAttribute('data-ad-unit-path', adUnitPath);
  conditionalAdElem.className = 'extraAd';
  return conditionalAdElem;
};

export const isHidden = (el) =>
  !(el.offsetWidth || el.offsetHeight || el.getClientRects().length);

export const getSizeFromAttributes = (el, width, breakPoints) => {
  if (width >= breakPoints.laptop && el.dataset.desktopSizes) {
    return el.dataset.desktopSizes;
  }

  if (width >= breakPoints.tablet && el.dataset.tabletSizes) {
    return el.dataset.tabletSizes;
  }

  if (width < breakPoints.tablet && el.dataset.mobileSizes) {
    return el.dataset.mobileSizes;
  }

  if (el.dataset.sizes) {
    return el.dataset.sizes;
  }
};

export const findMPUsForRefreshCheck = (elem = document) =>
  elem.querySelectorAll('[data-mpu][data-refresh-check="true"][id]');

export const findUninitialisedMPUs = (elem = document) =>
  elem.querySelectorAll('[data-mpu]:not([data-initialised="true"])');

export const findUninitialisedVisibleMPUs = (elem = document) =>
  [...thisModule.findUninitialisedMPUs(elem)].filter(
    (mpu) => !thisModule.isHidden(mpu),
  );

export const getConfig = (mpuName, pageWidth, adTypes) => {
  const mpuSizeConfig = adTypes[mpuName];

  if (mpuSizeConfig) {
    const config = [...mpuSizeConfig]
      .reverse()
      .find((item) => item.from <= pageWidth);

    if (config) {
      return config;
    }
  }
};

export const getSizeFromConfig = (mpuName, pageWidth, adTypes) => {
  const config = getConfig(mpuName, pageWidth, adTypes);
  return config ? config.supportedSizes : null;
};

export const getSizeConfigAtWidth = (el, width, adTypes, breakPoints) => {
  if (el.dataset.sizeKey) {
    const sizeArr = getSizeFromConfig(el.dataset.sizeKey, width, adTypes);
    sizeArr ||
      console.warn(`Incorrect Ad size detected: ${el.dataset?.tileName}`);
    return sizeArr
      ? {
          str: sizeArr.map((size) => size.join('x')).join('|'),
          arr: sizeArr,
        }
      : null;
  }

  const sizeString = getSizeFromAttributes(el, width, breakPoints) || '300x250';

  return {
    str: sizeString,
    arr: sizeString
      .split('|')
      .map((size) => size.split('x').map((dimension) => Number(dimension))),
  };
};

export const getAdConfigAtWidth = (el, width, adTypes, breakPoints) => {
  return {
    supportedSizes: getSizeConfigAtWidth(el, width, adTypes, breakPoints),
    refreshMS: el.dataset.refreshMs,
  };
};

export const createSlot = (adUnitPath, size, id, targeting = {}) =>
  window.googletag
    .defineSlot(adUnitPath, size, id)
    .setCollapseEmptyDiv(true)
    .updateTargetingFromMap(targeting)
    .addService(window.googletag.pubads());

export const initialiseAd = (
  el,
  { registerAsCurrentAd, adConfig: { types }, breakPoints },
) => {
  const { supportedSizes, refreshMS } = getAdConfigAtWidth(
    el,
    getWindowWidth(),
    types,
    breakPoints,
  );

  const slot = supportedSizes?.arr?.length
    ? createSlot(el.dataset.adUnitPath, supportedSizes.arr, el.id, {
        tile: el.id,
      })
    : null;

  if (slot) {
    registerAsCurrentAd(el, slot, supportedSizes.str, refreshMS);
  }

  return !!slot;
};

export const destroyConditionalAds = (slotIds, destroyAdFn) => {
  [...document.getElementsByClassName('extraAd')]?.forEach((el) => {
    const idParts = el.id.split(el.dataset.tileNameExtra);
    const originatingSlotId = idParts
      .slice(0, -1)
      .join(el.dataset.tileNameExtra);

    if (!slotIds || slotIds.includes(originatingSlotId)) {
      destroyAdFn(el);
      el.remove();
    }
  });
};

export const destroyConditionalAdsOnRefresh = (destroyAd) => {
  window.googletag.cmd.push(() => {
    const originalRefresh = window.googletag.pubads().refresh;
    window.googletag.pubads().refresh = (slots, ...otherArgs) => {
      destroyConditionalAds(
        slots?.map((slot) => slot?.getSlotElementId()).filter((x) => !!x),
        destroyAd,
      );
      originalRefresh(slots, ...otherArgs);
    };
  });
};

export const getAdRefreshIndex = (el) =>
  Number(document.getElementById(`${el.id}_adIndex`).innerHTML);

export const createAdUnitPathFromOther = (adUnitPath, adId) =>
  [...adUnitPath.split('/').slice(0, -1), adId].join('/');

export const loadOutOfPageAd = () => {
  const outOfPageAdId = 'outofpageslot';
  const existingAd = document.querySelector('[data-mpu]');

  // only load if there are other ads on page
  if (existingAd) {
    window.googletag
      .defineOutOfPageSlot(
        createAdUnitPathFromOther(existingAd.dataset.adUnitPath, outOfPageAdId),
        outOfPageAdId,
      )
      .addService(window.googletag.pubads());
  }
};

export const refreshDetectedAdMetaData = (detectedAds) => {
  const duplicateLineItemCounts = Object.values(detectedAds).reduce(
    (acc, item) => ({
      ...acc,
      [item.lineItemId]: (acc[item.lineItemId] || 0) + 1,
    }),
    {},
  );

  Object.entries(detectedAds).forEach(([id, details]) => {
    detectedAds[id].lineItemCount = duplicateLineItemCounts[details.lineItemId];
  });
};
