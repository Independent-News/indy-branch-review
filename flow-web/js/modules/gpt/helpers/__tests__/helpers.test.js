/**
 * @jest-environment jsdom
 */
import { getWindowWidth } from '#app/public/js/modules/util';

import * as helpersModule from '../helpers';

const {
  createAd,
  isHidden,
  getSizeFromAttributes,
  findMPUsForRefreshCheck,
  findUninitialisedMPUs,
  getConfig,
  getSizeFromConfig,
  getSizeConfigAtWidth,
} = helpersModule;

describe('gpt/helpers.js', () => {
  describe('createAd', () => {
    it('creates an ad element', () => {
      const adElement = createAd('AD_ID', '100x100', '/some/ad/unit/path');

      expect(adElement).toMatchSnapshot();
    });
  });

  describe('isHidden', () => {
    let testElem;
    let mockClientRects;

    beforeEach(() => {
      mockClientRects = [];
      testElem = {
        offsetHeight: 0,
        offsetWidth: 0,
        getClientRects: () => mockClientRects,
      };
    });

    it('will return false if element is visible on page', () => {
      testElem.offsetHeight = 100;

      expect(isHidden(testElem)).toBeFalsy();

      testElem.offsetHeight = 0;
      testElem.offsetWidth = 100;

      expect(isHidden(testElem)).toBeFalsy();

      testElem.offsetWidth = 0;
      mockClientRects = [{}];

      expect(isHidden(testElem)).toBeFalsy();

      testElem.offsetHeight = 10;
      testElem.offsetWidth = 10;
      mockClientRects = [{}];

      expect(isHidden(testElem)).toBeFalsy();
    });

    it('will return false if element is not visible on page', () => {
      expect(isHidden(testElem)).toBeTruthy();
    });
  });

  describe('getSizeFromAttributes(el, width, breakPoints)', () => {
    const breakPoints = { laptop: 1000, tablet: 500 };
    const elem = document.createElement('div');
    elem.dataset.desktopSizes = 'DESKTOP_SIZES';
    elem.dataset.tabletSizes = 'TABLET_SIZES';
    elem.dataset.mobileSizes = 'MOBILE_SIZES';

    it('will report desktop sizes if desktop sized', () => {
      expect(getSizeFromAttributes(elem, 1500, breakPoints)).toBe(
        'DESKTOP_SIZES',
      );
    });

    it('will report tablet sizes if tablet sized', () => {
      expect(getSizeFromAttributes(elem, 700, breakPoints)).toBe(
        'TABLET_SIZES',
      );
    });

    it('will report mobile sizes if mobile sized', () => {
      expect(getSizeFromAttributes(elem, 300, breakPoints)).toBe(
        'MOBILE_SIZES',
      );
    });

    it("will report device size independent sizes if general 'size' attribute is specified alone", () => {
      const elem = document.createElement('div');
      elem.dataset.sizes = 'SOME_SIZES';

      expect(getSizeFromAttributes(elem, 300, breakPoints)).toBe('SOME_SIZES');
    });
  });

  describe('findMPUsForRefeshCheck(elem)', () => {
    it('will find all MPUs that require a refresh check', () => {
      const testElem = document.createElement('div');

      testElem.innerHTML = `
<div id="mpu1" data-refresh-check="true" data-mpu></div>
<div id="mpu2" data-refresh-check="false" data-mpu></div>
<div id="mpu3" data-mpu></div>
<div></div>
<div id="mpu4" data-refresh-check="true" data-mpu></div>
<div id="mpu5" data-mpu></div>
<div id="mpu6" data-refresh-check="true" data-mpu></div>
<div id="mpu7" data-refresh-check="true"></div>`;

      expect(findMPUsForRefreshCheck(testElem)).toMatchSnapshot();
    });
  });

  describe('findUninitialisedMPUs(elem)', () => {
    it('can find uninitialised MPUs', () => {
      const testElem = document.createElement('div');

      testElem.innerHTML = `
<div id="mpu1" data-mpu></div>
<div id="mpu2" data-initialised="true" data-mpu></div>
<div id="mpu3" data-mpu></div>
<div></div>
<div id="mpu4" data-initialised="false" data-mpu></div>
<div id="mpu5" data-mpu></div>
<div id="mpu6" data-mpu></div>
<div id="mpu7"></div>`;

      expect(findUninitialisedMPUs(testElem)).toMatchSnapshot();
    });
  });

  describe('findUninitialisedVisibleMPUs(elem)', () => {
    const mockParentElem = document.createElement('div');

    mockParentElem.innerHTML = `
<div id="mpu1"></div>
<div id="mpu2"></div>
<div id="mpu3"></div>
<div id="mpu4"></div>
<div id="mpu5"></div>`;

    beforeAll(async () => {
      jest
        .spyOn(helpersModule, 'findUninitialisedMPUs')
        .mockReturnValue(mockParentElem.children);
      jest
        .spyOn(helpersModule, 'isHidden')
        .mockReturnValueOnce(true)
        .mockReturnValueOnce(false)
        .mockReturnValueOnce(true)
        .mockReturnValue(false);
    });

    it('can find uninitialised visible MPUs', () => {
      expect(
        helpersModule.findUninitialisedVisibleMPUs(mockParentElem),
      ).toMatchSnapshot();

      expect(helpersModule.findUninitialisedMPUs).toBeCalledTimes(1);
      expect(helpersModule.findUninitialisedMPUs).toHaveBeenCalledWith(
        mockParentElem,
      );
      expect(helpersModule.isHidden).toBeCalledTimes(5);
    });
  });

  describe('getConfig(mpuName, pageWidth, adConfig)', () => {
    const mockAdTypes = {
      AN_MPU_NAME: [
        { from: 0, ident: 'mobile' },
        { from: 600, ident: 'tablet' },
        { from: 1000, ident: 'desktop' },
      ],
    };

    it('will get the correct config for an ad based on page width', () => {
      expect(getConfig('AN_MPU_NAME', 340, mockAdTypes).ident).toBe('mobile');
      expect(getConfig('AN_MPU_NAME', 700, mockAdTypes).ident).toBe('tablet');
      expect(getConfig('AN_MPU_NAME', 999, mockAdTypes).ident).toBe('tablet');
      expect(getConfig('AN_MPU_NAME', 1000, mockAdTypes).ident).toBe('desktop');
      expect(getConfig('AN_MPU_NAME', Infinity, mockAdTypes).ident).toBe(
        'desktop',
      );
    });

    it('will return undefined if MPU name is not recognised', () => {
      expect(getConfig('UNKNOWN_MPU', Infinity, mockAdTypes)).toBeUndefined();
    });
  });

  describe('getSizeFromConfig(mpuName, pageWidth, adConfig)', () => {
    const mockAdTypes = {
      AN_MPU_NAME: [
        { from: 600, ident: 'tablet' },
        {
          from: 1000,
          ident: 'desktop',
          supportedSizes: [
            [3, 2],
            [4, 3],
          ],
        },
      ],
    };

    it('will get size from config', () => {
      expect(getSizeFromConfig('AN_MPU_NAME', Infinity, mockAdTypes)).toEqual([
        [3, 2],
        [4, 3],
      ]);
    });

    it('will return null if MPU is not recognised', () => {
      expect(getSizeFromConfig('UNKNOWN_MPU', 1100, mockAdTypes)).toBeNull();
    });

    it('will return null if size is not supported', () => {
      expect(getSizeFromConfig('AN_MPU_NAME', 200, mockAdTypes)).toBeNull();
    });
  });

  describe('getAdSizes(el, width, adConfig, breakPoints)', () => {
    let adSizes;

    const elm = document.createElement('div');
    const breakPoints = { desktop: 1000, tablet: 600 };

    const mockAdTypes = {
      SKIN: [
        {
          from: 1000,
          supportedSizes: [
            [300, 250],
            [300, 600],
          ],
        },
      ],
    };

    describe('given elm is using a sizeKey dataset', () => {
      beforeAll(() => {
        elm.dataset.sizeKey = 'SKIN';

        adSizes = getSizeConfigAtWidth(
          elm,
          getWindowWidth(),
          mockAdTypes,
          breakPoints,
        );
      });

      afterAll(() => {
        elm.dataset.sizeKey = '';
      });

      it('will have a valid sizeKey value', () => {
        expect(elm.dataset.sizeKey).toBeTruthy();
      });

      it('will return ad sizes', () => {
        expect(adSizes).toEqual({
          str: '300x250|300x600',
          arr: [
            [300, 250],
            [300, 600],
          ],
        });
      });

      it('will return null for unsupported sizes', () => {
        const sizes = getSizeConfigAtWidth(elm, 500, mockAdTypes, breakPoints);

        expect(sizes).toBeNull();
      });
    });

    describe('given elm is not using a sizeKey dataset', () => {
      beforeAll(() => {
        adSizes = getSizeConfigAtWidth(
          elm,
          getWindowWidth(),
          mockAdTypes,
          breakPoints,
        );
      });

      it('will not have a sizeKey value', () => {
        expect(elm.dataset.sizeKey).toBeFalsy();
      });

      it('will return size', () => {
        expect(adSizes).toEqual({
          str: '300x250',
          arr: [[300, 250]],
        });
      });
    });
  });

  describe('createSlot(adUnitPath, size, id, targeting = {})', () => {});

  describe('initialiseAd(el, { registerAsCurrentAd, adConfig, breakPoints })', () => {});

  describe('destroyConditionalAds(slotIds, destroyAdFn)', () => {});

  describe('destroyConditionalAdsOnRefresh(destroyAd)', () => {});

  describe('getAdRefreshIndex(el)', () => {});

  describe('createAdUnitPathFromOther(adUnitPath, adId)', () => {});

  describe('loadOutOfPageAd()', () => {});

  describe('refreshDetectedAdsMetaData', () => {
    it('should add the total current detected ads that share the same lineItemId', () => {
      const dummyData = {
        bish: { lineItemId: 'xxx' },
        bash: { lineItemId: 'xxx' },
        fizz: { lineItemId: 'xxx' },
        buzz: { lineItemId: 'yyy' },
      };

      helpersModule.refreshDetectedAdMetaData(dummyData);

      expect(dummyData).toEqual({
        bish: { lineItemId: 'xxx', lineItemCount: 3 },
        bash: { lineItemId: 'xxx', lineItemCount: 3 },
        fizz: { lineItemId: 'xxx', lineItemCount: 3 },
        buzz: { lineItemId: 'yyy', lineItemCount: 1 },
      });
    });
  });
});
