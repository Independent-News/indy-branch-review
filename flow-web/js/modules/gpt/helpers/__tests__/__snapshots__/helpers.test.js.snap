// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`gpt/helpers.js createAd creates an ad element 1`] = `
<div
  class="extraAd"
  data-ad-unit-path="/some/ad/unit/path"
  data-mpu="true"
  data-size="100x100"
  id="AD_ID"
/>
`;

exports[`gpt/helpers.js findMPUsForRefeshCheck(elem) will find all MPUs that require a refresh check 1`] = `
NodeList [
  <div
    data-mpu=""
    data-refresh-check="true"
    id="mpu1"
  />,
  <div
    data-mpu=""
    data-refresh-check="true"
    id="mpu4"
  />,
  <div
    data-mpu=""
    data-refresh-check="true"
    id="mpu6"
  />,
]
`;

exports[`gpt/helpers.js findUninitialisedMPUs(elem) can find uninitialised MPUs 1`] = `
NodeList [
  <div
    data-mpu=""
    id="mpu1"
  />,
  <div
    data-mpu=""
    id="mpu3"
  />,
  <div
    data-initialised="false"
    data-mpu=""
    id="mpu4"
  />,
  <div
    data-mpu=""
    id="mpu5"
  />,
  <div
    data-mpu=""
    id="mpu6"
  />,
]
`;

exports[`gpt/helpers.js findUninitialisedVisibleMPUs(elem) can find uninitialised visible MPUs 1`] = `
[
  <div
    id="mpu2"
  />,
  <div
    id="mpu4"
  />,
  <div
    id="mpu5"
  />,
]
`;
